src/Tools/ExternalAccess/LiveShare/RemoteLanguageServiceWorkspace.cs

```diff
@@ -0,0 +1,444 @@
+        {
+            if (_openedDocs.ContainsKey(moniker))
+            {
+                return;
+            }
+
+            Document document = GetOrAddDocument(moniker);
+
+            if (document != null)
+            {
+                SourceTextContainer textContainer = textBuffer.AsTextContainer();
+                OnDocumentOpened(document.Id, textContainer);
+                _openedDocs = _openedDocs.SetItem(moniker, document.Id);
+            }
+        }
+
+        private bool IsExternalLocalUri(string localPath)
+        {
+            return localPath.StartsWith(_externalPath) &&
+                localPath.Length > (_externalPath.Length + 1);
+        }
+
+        public Document GetOrAddDocument(string filePath)
+        {
+            DocumentId docId = CurrentSolution.GetDocumentIdsWithFilePath(filePath).FirstOrDefault();
+            if (docId != null)
+            {
+                return CurrentSolution.GetDocument(docId);
+            }
+
+            if (!IsRemoteSession)
+            {
+                return null;
+            }
+
+            // If the document is within the joined folder or it's a registered external file,
+            // add it to the workspace, otherwise bail out.
+            if (!filePath.StartsWith(_remoteRootPath) &&
+                !IsExternalLocalUri(filePath))
+            {
+                return null;
+            }
+
+            var language = GetLanguage(filePath);
+            // Unsupported language.
+            if (language == null)
+            {
+                return null;
+            }
+
+            var folderName = Path.GetFileNameWithoutExtension(_remoteRootPath);
+            return AddDocumentToProject(filePath, language, folderName);
+        }
+
+        public Document GetOrAddExternalDocument(string filePath, string language)
+        {
+            DocumentId docId = CurrentSolution.GetDocumentIdsWithFilePath(filePath).FirstOrDefault();
+            if (docId != null)
+            {
+                return CurrentSolution.GetDocument(docId);
+            }
+
+            return AddDocumentToProject(filePath, language, ExternalProjectName);
+        }
+
+        public async Task<DocumentSpan?> GetDocumentSpanFromLocation(LSP.Location location, CancellationToken cancellationToken)
+        {
+            var document = GetOrAddDocument(location.Uri.LocalPath);
+            if (document == null)
+            {
+                return null;
+            }
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // The protocol converter would have synced the file to disk but we the document snapshot that was in the workspace before the sync would have empty text.
+            // So we need to read from disk in order to map from line\column to a textspan.
+            if (string.IsNullOrEmpty(text.ToString()))
```

Comment:
you can just check length rather than doing ToString which could expensive for case where content is already there.

by the way, I dont think document.GetTextAsyn can ever return null?